### PR TITLE
Use views instead of dto

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
This PR is a major refactoring to separate server-side code from client-side code:

-Client side now only use Views for the roster and rotation screen
-Rotation API changed; now you send/receive a rotation view, which includes rotationLength, spotList, employeeList and spotIdToShiftTemplateViewListMap, so you don't need to get them in seperate REST calls
-View use LocalDateTime instead of OffsetDateTime; Server still use OffsetDateTime
-Created an interface `HasTImeslot` that ShiftView, EmployeeAvaliabilityView and ShiftTemplateView all implement, and created a generic Blob type that uses HasTimeslot to avoid duplicating functionality.

### CODE REVIEW

**Comment by Christopher-Chianelli** [comment]
-Also edited `Duration` of super-sources so it is mostly compatiable with `java.time.Duration`

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/employeeroster/EmployeeBlob.java [hunk 1]
```diff
@@ -16,143 +16,103 @@
...
 
-    private OffsetDateTime getStartDateTime() {
-        if (null != shift) {
-            return shift.getStartDateTime();
+    public String getLabel() {
+        if (shiftView != null) {
+            return getSpot().toString() + ": " + shiftView.getStartDateTime() + "-" + shiftView.getEndDateTime();
         } else {
-            return employeeAvailability.getStartDateTime();
+            return employeeAvailabilityView.getState().toString() + ":" + employeeAvailabilityView.getStartDateTime() + "-" + employeeAvailabilityView.getEndDateTime();
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/employeeroster/EmployeeBlob.java]: on enum's, use name() instead of toString() if it's for the UI

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/employeeroster/EmployeeBlob.java [hunk 2]
```diff
@@ -16,143 +16,103 @@
...
-
-    public void setEmployeeAvailability(final EmployeeAvailability employeeAvailability) {
-        this.employeeAvailability = employeeAvailability;
     }
 
     @Override
-    public LinearScale<OffsetDateTime> getScale() {
-        return scale;
+    public HasTimeslot getTimeslot() {
+        // TODO Auto-generated method stub
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/employeeroster/EmployeeBlob.java]: todo ^

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/rotation/RotationPage.java [hunk 3]
```diff
@@ -212,29 +134,29 @@ private void onRefreshClicked(final @ForEvent("click") MouseEvent e) {
...
+                                    if (!shiftTemplateBlob.getTwin().isPresent() ||
+                                            shiftTemplateBlob.getTwin().get().getPositionInGridPixels() < shiftTemplateBlob.getPositionInGridPixels()) {
+                                        shiftTemplates.add(shiftTemplateBlob.getShiftTemplateView());
+                                    }
+                                });
+                            });
+                            return shiftTemplates;
+                        }));
+
+        RotationView out = new RotationView();
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/rotation/RotationPage.java]: can we rename this variable "out" to rotationView?

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/spotroster/ShiftBlobPopoverContent.java [hunk 4]
```diff
@@ -122,35 +121,35 @@ public void init(final BlobView<?, ?> blobView) {
...
             employees.forEach(e -> employeeSelect.addItem(e.getName(), e.getId().toString()));
-            employeeSelect.setSelectedIndex((shift.getEmployee() == null) ? 0 : employees.indexOf(shift.getEmployee()) + 1);
+            employeeSelect.setSelectedIndex((blob.getEmployee() == null) ? 0 : employees.indexOf(blob.getEmployee()) + 1);
         }));
 
-        final OffsetDateTime start = shift.getStartDateTime();
+        final LocalDateTime start = shift.getStartDateTime();
         fromDay.value = start.getMonth().toString() + " " + start.getDayOfMonth(); //FIXME: i18n
-        fromHour.value = GwtJavaTimeWorkaroundUtil.toLocalTime(start) + "";
+        fromHour.value = start + "";
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/spotroster/ShiftBlobPopoverContent.java]: don't concatonate "" just to cast to a String. Use Integer.toString(start) instead.

File: optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/Duration.java [hunk 5]
```diff
@@ -295,67 +295,56 @@ public static Duration between(DateTimeAccessor startInclusive, DateTimeAccessor
...
-        case 8:
-          nanos *= 10;
-          break;
-      }
-      return negative ? ofSeconds(secs, -nanos) : create(secs, nanos);
-
+        Duration out = Duration.ZERO;
+        StringBuilder number = new StringBuilder();
+        int state = 0;
+        for (int i = 2; i < text.length(); i++) {
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/Duration.java]: Where do these changes come from?

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/employeeroster/EmployeeBlob.java (hunk 1 above)
Comment (Christopher-Chianelli) [reply optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/pages/employeeroster/EmployeeBlob.java]: Javadoc says "Most programmers should use the toString() method in preference to this one, as the toString method may return a more user-friendly name" for the `name()` method, are you sure?

File: optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/LocalDateTime.java [hunk 6]
```diff
@@ -480,6 +480,10 @@ public long getLong(DateTimeField field) {
 
     if (field instanceof ChronoField) {
       ChronoField f = (ChronoField) field;
+      if (f.equals(ChronoField.INSTANT_SECONDS)) {
+          return this.time.getSecond() + this.time.getMinute() * 60 + this.time.getHour() * 60 * 60 +
+                  this.date.toEpochDay() * 60 * 60 * 24;
+      }
```
Comment (ge0ffrey) [inline optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/LocalDateTime.java]: And these, where do they come from?

File: optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/shift/ShiftRestServiceImpl.java [hunk 7]
```diff
@@ -115,18 +130,60 @@ public Boolean removeShift(Integer tenantId, Long id) {
...
     @Override
-    public List<ShiftTemplate> getShiftTemplateList(Integer tenantId) {
-        TypedQuery<ShiftTemplate> q = entityManager.createNamedQuery("ShiftTemplate.findAll", ShiftTemplate.class);
-        q.setParameter("tenantId", tenantId);
-        return q.getResultList();
+    public RotationView getRotation(Integer tenantId) {
+        List<ShiftTemplate> shiftTemplateList = entityManager.createNamedQuery("ShiftTemplate.findAll", ShiftTemplate.class)
+                .setParameter("tenantId", tenantId)
+                .getResultList();
+        RotationView out = new RotationView();
```
Comment (ge0ffrey) [inline optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/shift/ShiftRestServiceImpl.java]: don't name it out

File: optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/shift/ShiftRestServiceImpl.java [hunk 8]
```diff
@@ -115,18 +130,60 @@ public Boolean removeShift(Integer tenantId, Long id) {
...
+        });
+        out.setSpotIdToShiftTemplateViewListMap(spotIdToShiftTemplateViewListMap);
+        return out;
     }
 
     @Override
     @Transactional
-    public void updateShiftTemplate(Integer tenantId, List<ShiftTemplate> shiftTemplateList) {
-        List<ShiftTemplate> oldShiftTemplateList = getShiftTemplateList(tenantId);
+    public void updateRotation(Integer tenantId, RotationView rotationView) {
```
Comment (ge0ffrey) [inline optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/shift/ShiftRestServiceImpl.java]: This update the entire rotation in batch?

File: optashift-employee-rostering-shared/src/main/java/org/optaplanner/openshift/employeerostering/shared/roster/RosterState.java [hunk 9]
```diff
@@ -63,15 +65,30 @@ public boolean isDraft(OffsetDateTime dateTime) {
     }
 
     @JsonIgnore
-    public boolean isHistoric(Shift shift) {
-        return isHistoric(shift.getStartDateTime());
+    public boolean isPublished(OffsetDateTime dateTime) {
+        return !isHistoric(dateTime) && !isDraft(dateTime);
```
Comment (ge0ffrey) [inline optashift-employee-rostering-shared/src/main/java/org/optaplanner/openshift/employeerostering/shared/roster/RosterState.java]: isHistoric must be before isDraft which must be before isPublished.
Please revert these changes.

Comment (ge0ffrey) [reply optashift-employee-rostering-shared/src/main/java/org/optaplanner/openshift/employeerostering/shared/roster/RosterState.java]: ah, I get what happened. Just move isPublished(OffsetDatetime) before or after the other isPublished(LocalDateTime) method

File: optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/Duration.java (hunk 5 above)
Comment (Christopher-Chianelli) [reply optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/Duration.java]: `java.time.Duration` toString() (sent by server) may have hours and minutes, the old implementation only accepted seconds

File: optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/LocalDateTime.java (hunk 6 above)
Comment (Christopher-Chianelli) [reply optashift-employee-rostering-gwtui/src/main/super/org/optaplanner/openshift/employeerostering/gwtui/emulate/java/time/LocalDateTime.java]: The ability to directly add a `Duration` to `LocalDateTime`; mostly for my sanity

File: optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/shift/ShiftRestServiceImpl.java (hunk 8 above)
Comment (Christopher-Chianelli) [reply optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/shift/ShiftRestServiceImpl.java]: Yes

File: optashift-employee-rostering-shared/src/test/java/org/optaplanner/openshift/employeerostering/shared/rotation/ShiftTemplateTest.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Test;
+import org.optaplanner.openshift.employeerostering.shared.employee.Employee;
+import org.optaplanner.openshift.employeerostering.shared.rotation.view.ShiftTemplateView;
+import org.optaplanner.openshift.employeerostering.shared.spot.Spot;
+
+import static org.junit.Assert.assertEquals;
+
+public class ShiftTemplateTest {
+
+    final int ROTATION_LENGTH = 7;
```
Comment (ge0ffrey) [inline optashift-employee-rostering-shared/src/test/java/org/optaplanner/openshift/employeerostering/shared/rotation/ShiftTemplateTest.java]: private

**Comment by Christopher-Chianelli** [comment]
jenkins please retest this
